File: planet-examples/src/main/java/org/saar/example/gui/GuiExample.java
package org.saar.example.gui;

import org.saar.core.renderer.RenderContextBase;
import org.saar.gui.UIComponent;
import org.saar.gui.UIContainer;
import org.saar.gui.UIDisplay;
import org.saar.gui.component.UIButton;
import org.saar.gui.component.UICheckbox;
import org.saar.gui.component.UISlider;
import org.saar.gui.style.Colour;
import org.saar.gui.style.value.CoordinateValues;
import org.saar.gui.style.value.LengthValues;
import org.saar.lwjgl.glfw.input.keyboard.Keyboard;
import org.saar.lwjgl.glfw.window.Window;
import org.saar.lwjgl.opengl.utils.GlBuffer;
import org.saar.lwjgl.opengl.utils.GlUtils;

public class GuiExample {

    private static final int WIDTH = 700;
    private static final int HEIGHT = 500;

    public static void main(String[] args) {
        final Window window = Window.create("Lwjgl", WIDTH, HEIGHT, true);

        final UIDisplay display = new UIDisplay(window);

        final UIContainer uiContainer = new UIContainer();

        final UISlider sizeUiSlider = new UISlider();
        sizeUiSlider.getStyle().getY().set(450);
        sizeUiSlider.getStyle().getX().set(CoordinateValues.center());
        sizeUiSlider.getStyle().getWidth().set(LengthValues.percent(90));
        sizeUiSlider.getStyle().getHeight().set(20);

        sizeUiSlider.getDynamicValueProperty().addListener(e -> {
            final float percents = (e.getNewValue().floatValue() / 100) * 50 + 20;
            uiContainer.getStyle().getWidth().set(LengthValues.percent(percents));
        });

        display.add(sizeUiSlider);


        uiContainer.getStyle().getWidth().set(LengthValues.percent(50));
        uiContainer.getStyle().getHeight().set(LengthValues.ratio(1));
        uiContainer.getStyle().getX().set(CoordinateValues.center());
        display.add(uiContainer);

        final UIComponent uiComponent = new MyUIComponent();
        uiContainer.add(uiComponent);

        final UIButton uiButton = new UIButton();
        uiButton.getStyle().getX().set(CoordinateValues.center());
        uiButton.getStyle().getY().set(CoordinateValues.center());
        uiButton.getStyle().getWidth().set(LengthValues.percent(10));
        uiButton.getStyle().getHeight().set(LengthValues.ratio(.5f));
        uiButton.setOnAction(e -> System.out.println("Clicked!"));
        uiContainer.add(uiButton);

        final UISlider uiSlider = new UISlider();
        uiSlider.getStyle().getY().set(20);
        uiSlider.getStyle().getX().set(CoordinateValues.center());
        uiSlider.getStyle().getWidth().set(LengthValues.percent(90));
        uiSlider.getStyle().getHeight().set(20);

        uiSlider.getDynamicValueProperty().addListener(e -> {
            final float percents = e.getNewValue().floatValue() / 2;
            uiButton.getStyle().getWidth().set(LengthValues.percent(percents));
        });

        uiContainer.add(uiSlider);

        final UICheckbox uiCheckbox = new UICheckbox();
        uiCheckbox.getStyle().getY().set(50);
        uiCheckbox.getStyle().getX().set(CoordinateValues.percentCenter(20));
        uiCheckbox.getStyle().getWidth().set(LengthValues.pixels(20));
        uiCheckbox.getStyle().getBackgroundColour().set(new Colour(48, 63, 159, 1f));
        uiCheckbox.getStyle().getRadiuses().set(3);
        uiContainer.add(uiCheckbox);

        final Keyboard keyboard = window.getKeyboard();
        while (window.isOpen() && !keyboard.isKeyPressed('E')) {
            GlUtils.clear(GlBuffer.COLOUR);
            display.render(new RenderContextBase(null));

            display.update();
            window.update(true);
            window.pollEvents();
        }

        display.delete();
        window.destroy();
    }

}

File: planet-examples/src/main/java/org/saar/example/gui/GuiExample.kt
package org.saar.example.gui

import org.jproperty.Observable
import org.lwjgl.glfw.GLFW
import org.saar.core.renderer.RenderContext
import org.saar.gui.UIDisplay
import org.saar.gui.UIElement
import org.saar.gui.component.UIButton
import org.saar.gui.component.UICheckbox
import org.saar.gui.component.UISlider
import org.saar.gui.event.EventListener
import org.saar.gui.style.alignment.AlignmentValues.vertical
import org.saar.gui.style.arrangement.ArrangementValues.spaceBetween
import org.saar.gui.style.axisalignment.AxisAlignmentValues.center
import org.saar.gui.style.length.LengthValues.fill
import org.saar.gui.style.length.LengthValues.percent
import org.saar.gui.style.length.LengthValues.pixels
import org.saar.gui.style.length.LengthValues.ratio
import org.saar.lwjgl.glfw.window.Window
import org.saar.lwjgl.opengl.utils.GlBuffer
import org.saar.lwjgl.opengl.utils.GlUtils

private const val WIDTH = 700
private const val HEIGHT = 500

fun main() {
    val window = Window.create("Lwjgl", WIDTH, HEIGHT, true)

    val display = UIDisplay(window) {
        style.alignment.value = vertical
        style.arrangement.value = spaceBetween
        style.axisAlignment.value = center

        val sizeUiSlider = +UISlider {
            style.width.value = fill
            style.height.value = pixels(30)
            style.margin.set(20)
        }

        val uiContainer = +UIElement {
            style.width.value = percent(50f)
            style.alignment.value = vertical
            style.axisAlignment.value = center

            +MyUIComponent()

            val uiButton = +UIButton {
                style.width.value = percent(10f)
                style.height.value = ratio(.5f)
                onAction = EventListener { println("Clicked!") }
            }

            +UISlider {
                style.width.value = percent(90f)
                style.height.value = pixels(40)

                dynamicValueProperty.addListener { _: Observable ->
                    val percents = dynamicValueProperty.floatValue / 2
                    uiButton.style.width.value = percent(percents)
                }
            }

            +UICheckbox {
                style.width.value = pixels(30)
            }
        }

        sizeUiSlider.dynamicValueProperty.addListener { _: Observable ->
            val percents = sizeUiSlider.dynamicValueProperty.floatValue / 100 * 50 + 20
            uiContainer.style.width.value = percent(percents)
        }
    }

    while (window.isOpen && !window.keyboard.isKeyPressed(GLFW.GLFW_KEY_ESCAPE)) {
        GlUtils.setViewport(0, 0, window.width, window.height)
        GlUtils.clear(GlBuffer.COLOR)

        display.render(RenderContext())

        window.swapBuffers()
        window.pollEvents()
    }

    display.delete()
    window.destroy()
}